core/src/main/java/com/yahoo/imapnio/async/data/MessageNumberSet.java

```diff
@@ -0,0 +1,239 @@
+        this(start, start, SequenceType.LAST_MESSAGE_END);
+    }
+
+    /**
+     * Instantiates a sequence that only returns last message, aka, * .
+     *
+     * @param lastMsgFlag enum to denote whether it is a last message, should always be
+     * @throws ImapAsyncClientException when given isLastMessageOnly is false
+     */
+    public MessageNumberSet(@Nonnull final LastMessage lastMsgFlag) throws ImapAsyncClientException {
+        if (lastMsgFlag != LastMessage.LAST_MESSAGE) {
+            throw new ImapAsyncClientException(ImapAsyncClientException.FailureType.INVALID_INPUT);
+        }
+        this.start = -1;
+        this.end = -1;
+        this.seqType = SequenceType.LAST_MESSAGE_ONLY;
+    }
+
+    /**
+     * Instantiates a {@code MessageNumberSet} with start value, end value and SequenceType option.
+     *
+     * @param start starting message sequence or UID sequence
+     * @param end ending message sequence or UID sequence
+     * @param seqType ending option either a specific/absolute value, or last message as the end
+     */
+    private MessageNumberSet(final long start, final long end, @Nonnull final SequenceType seqType) {
+        this.start = start;
+        this.end = end;
+        this.seqType = seqType;
+    }
+
+    /**
+     * Converts an array of integers into an array of MessageNumberSet. This is a helper method for callers with int array.
+     *
+     * @param msgs array of primitive integer data type message number (could be message sequence or UID)
+     * @return MessageNumberSet array
+     */
+    public static MessageNumberSet[] createMessageNumberSets(@Nonnull final int[] msgs) {
+        List<MessageNumberSet> v = new ArrayList<MessageNumberSet>();
+        int i, j;
+        for (i = 0; i < msgs.length; i++) {
+            int start = msgs[i];
+
+            // Look for contiguous elements
+            for (j = i + 1; j < msgs.length; j++) {
+                if (msgs[j] != msgs[j - 1] + 1) {
+                    break;
+                }
+            }
+            int end = msgs[j - 1];
+            v.add(new MessageNumberSet(start, end));
+            i = j - 1; // i gets incremented @ top of the loop
+        }
+        return v.toArray(new MessageNumberSet[v.size()]);
+    }
+
+    /**
+     * Converts an array of long into array of MessageNumberSet. This is a helper method for callers with long array.
+     *
+     * @param msgs array of long data type message number (could be message sequence or UID)
+     * @return the string generated that conforms to RFC3501 Message Number syntax
+     */
+    public static MessageNumberSet[] createMessageNumberSets(@Nonnull final long[] msgs) {
+        List<MessageNumberSet> v = new ArrayList<MessageNumberSet>();
+        int i, j;
+        for (i = 0; i < msgs.length; i++) {
+            long start = msgs[i];
+
+            // Look for contiguous elements
+            for (j = i + 1; j < msgs.length; j++) {
```

Comment:
Would these look better as a while loop?

Reply:
Done.

Reply:
yours is better.  I changed.